conbench/buildinfo.py

```diff
@@ -0,0 +1,37 @@
+import json
+import logging
+from dataclasses import dataclass
+from typing import Optional
+
+import dacite
+
+log = logging.getLogger(__name__)
+
+
+@dataclass
+class Buildinfo:
+    # Full-length git commit hash.
+    commit: str
+    # Name of the branch carrying the commit.
+    branch_name: str
+    # Example value: 2023-02-06 12:01:01+00:00
+    build_time_rfc3339: str
+    build_hostname: str
+    # Canonical
+    version_string: str
+
+
+BUILD_INFO: Optional[Buildinfo] = None
+try:
+    # Try to discover additional build information at well-known path. For now,
+    # do not crash the application upon: file not found, file having unexpected
+    # contents, etc.
+    with open("/buildinfo.json", "rb") as fh:
+        data = json.loads(fh.read().decode("utf-8"))
```

Comment:
Any reason not to use `load`?
```suggestion
    with open("/buildinfo.json", "r") as fh:
        data = json.load(f)
```

Reply:
Oh. Thanks for asking! I could choose to not care, we could do that here, and it's unlikely that we would ever note the difference. That is true.

But this topic is important to me, and you asked for a reason why I wrote the code the way I wrote it. There is a pretty strong reason.

This is a principle I apply. "my version" is less complex and gives a useful error messages when it fails.

Opening files in "text mode" in Python uses an unbelievable amount of magic under the hood. This magic makes use of operating system features/environment specifics (locale). Magic that's great for the unaware developers that quickly want to get something working (I mean that!) -- but for thorough engineering and smooth maintenance it gets in the way, because it increases complexity, and wastes time when things go pear-shaped.

The following changelog entry shows that this topic is far from trivial and undergoes evolution -- `json.load(f)` can (since CPython 3.6) _also_ receive file objects that were opened with 'rb', and then the json module attempts to perform codec guessing:

> Changed in version 3.6: fp can now be a [binary file](https://docs.python.org/3/glossary.html#term-binary-file). The input encoding should be UTF-8, UTF-16 or UTF-32.

(When you do `open(path, 'r') as f: json.load(f)` you get a different codec guessing magic applied than as if you were doing `open(path, 'rb') as f: json.load(f)`.)

It is almost always better to be explicit about assumptions.  As a software engineer I can take burden from my future self and also fellow devs when I leave an explicit note about the character encoding that my code expects. That is why I typically do `rb` (give me the bytes!) and then specify the codec by name (describing the exact table by which I want bytes transformed into code points).

What's the magic that I don't want to have? From the Python docs, first a general statement (emphasis mine):

> Normally, files are opened in text mode, that means, you read and write strings from and to the file, which are encoded in a specific encoding. If encoding is not specified, **the default is platform dependent** (see [open()](https://docs.python.org/3/library/functions.html#open)). Because UTF-8 is the modern de-facto standard, **encoding="utf-8" is recommended unless you know that you need to use a different encoding**.

(https://docs.python.org/3/tutorial/inputoutput.html#reading-and-writing-files)

The docs for open() say:

> encoding is the name of the encoding used to decode or encode the file. This should only be used in text mode. **The default encoding is platform dependent** (whatever [locale.getencoding()](https://docs.python.org/3/library/locale.html#locale.getencoding) returns), but any [text encoding](https://docs.python.org/3/glossary.html#term-text-encoding) supported by Python can be used.

(https://docs.python.org/3/library/functions.html#open)

And now what even I find impressive now after having worked with CPython rather intensely for ~15 years, locale.getencoding() is _new_, it's documented with:

> Get the current [locale encoding](https://docs.python.org/3/glossary.html#term-locale-encoding): On Android and VxWorks, return "utf-8". On Unix, return the encoding of the current [LC_CTYPE] https://docs.python.org/3/library/locale.html#locale.LC_CTYPE) locale. Return "utf-8" if nl_langinfo(CODESET) returns an empty string: for example, if the current LC_CTYPE locale is not supported. On Windows, return the ANSI code page. The [Python preinitialization](https://docs.python.org/3/c-api/init_config.html#c-preinit) configures the LC_CTYPE locale. See also the [filesystem encoding and error handler](https://docs.python.org/3/glossary.html#term-filesystem-encoding-and-error-handler). This function is similar to [getpreferredencoding(False)](https://docs.python.org/3/library/locale.html#locale.getpreferredencoding) except this function ignores the [Python UTF-8 Mode](https://docs.python.org/3/library/os.html#utf8-mode). **New in version 3.11.**

(https://docs.python.org/3/library/locale.html#locale.getencoding)

"text mode" files in Python also have tremendous complexity with respect to newline-representing byte sequences.

Another wall of text :scroll:  It's just that I have spent a lot of time on the topic of character encoding in software before, in particular in Python:
- https://stackoverflow.com/a/17733233
- https://gehrcke.de/2015/12/how-to-raise-unicodedecodeerror-in-python-3/
- https://gehrcke.de/2014/07/mojibake-beatports-id3-text-encoding-is-broken/
- https://gehrcke.de/2014/02/command-line-argument-binary-data/

Reply:
Wow, thanks for the in-depth explanation! TIL a safer Python idiom.